cellmlmanip/model.py

```diff
@@ -436,18 +470,37 @@ def __connect(self, connection):
         """
         cmp_name_1, var_att_1, cmp_name_2, var_att_2 = self.__get_connection_endpoints(connection)
 
-        # Determine which are the source and target variables, and connect from->to
-        if (('public_interface', 'out') in var_att_1.items() or
-                ('private_interface', 'out') in var_att_1.items()) and (
-                    ('public_interface', 'in') in var_att_2.items() or
-                    ('private_interface', 'in') in var_att_2.items()):
-            return self.__connect_with_direction(cmp_name_1, var_att_1, cmp_name_2, var_att_2)
-        elif (('public_interface', 'out') in var_att_2.items() or
-              ('private_interface', 'out') in var_att_2.items()) and (
-                  ('public_interface', 'in') in var_att_1.items() or
-                  ('private_interface', 'in') in var_att_1.items()):
-            return self.__connect_with_direction(cmp_name_2, var_att_2, cmp_name_1, var_att_1)
-        raise RuntimeError("Cannot determine the source & target for connection %s" % connection)
+        # keys for lookup
+        pub = 'public_interface'
+        pri = 'private_interface'
+
+        def __are_siblings(name_a, name_b):
+            return self.components[name_a].parent == self.components[name_b].parent
+
+        def __parent_of(parent_name, child_name):
+            return parent_name == self.components[child_name].parent
+
+        def __has_interface(variable_a, key_a, value_a, variable_b, key_b, value_b):
+            return (key_a, value_a) in variable_a.items() and (key_b, value_b) in variable_b.items()
+
+        # if one component is encapsulated by the other
+        if __parent_of(cmp_name_1, cmp_name_2) or __parent_of(cmp_name_2, cmp_name_1):
+            # they can be connected with (i) public=out to private=in (ii) private=out to public=in
+            if __has_interface(*(var_att_1, pri, 'in'), *(var_att_2, pub, 'out')):
+                return self.__connect_with_direction(cmp_name_2, var_att_2, cmp_name_1, var_att_1)
+            elif __has_interface(*(var_att_1, pri, 'out'), *(var_att_2, pub, 'in')):
+                return self.__connect_with_direction(cmp_name_1, var_att_1, cmp_name_2, var_att_2)
+
+        # if components are siblings (in group or top-level group) or one encapsulates the other
```

Comment:
Again, think this is right

Reply:
Again, I don't think so! public-public only works for siblings, not encapsulation.

Reply:
```diff
--- i/tests/cellml_files/err_connect_to_hidden_component.cellml
+++ w/tests/cellml_files/err_connect_to_hidden_component.cellml
@@ -6,21 +6,30 @@


    <component name="parent">
+      <variable name="time2" public_interface="in" units="second"/>
    </component>

    <component name="child">
       <variable name="time" public_interface="out" units="second"/>
+      <variable name="time2" public_interface="out" units="second"/>
    </component>

    <component name="other">
       <variable name="time" public_interface="in" units="second"/>
    </component>

+   <!-- This isn't allowed as child is hidden from other -->
    <connection>
       <map_components component_1="child" component_2="other"/>
       <map_variables variable_1="time" variable_2="time"/>
    </connection>

+   <!-- This isn't allowed as the private_interface within parent is none -->
+   <connection>
+      <map_components component_1="child" component_2="parent"/>
+      <map_variables variable_1="time2" variable_2="time2"/>
+   </connection>
+
    <group>
       <relationship_ref relationship="encapsulation"/>
       <component_ref component="parent">
```

Reply:
Re. your second connection above, isn't that allowed according to spec: 

> Variables with a public_interface attribute value of "in" must be mapped to a single variable in a component _in the encapsulated_ or sibling sets with a public_interface attribute value of "out" or to a single variable in the parent of the current component with a private_interface attribute value of "out".

Reply:
Hmm, that contradicts other parts of the spec I think! Will check when I get a chance...

Reply:
If I understand correctly, what you're saying is that  (given a valid in/out pair) siblings are connected on the `public_interface` and parent/child are connected on `public`-`private` interfaces.

Reply:
From https://www.cellml.org/specifications/cellml_1.0/#sec_structure_spec_rules_map_variables_element :
>  A variable with a public_interface attribute value of "in" must be mapped to a single variable owned by a component in the sibling set, provided the target variable has a public_interface attribute value of "out", or to a single variable owned by the parent component, provided the target variable has a private_interface attribute value of "out".
>  A variable with a public_interface attribute value of "out" may be mapped to variables owned by components in the sibling set, provided the target variables have public_interface attribute values of "in". It may also be mapped to variables owned by the parent component, provided the target variables have private_interface attribute values of "in".
> A variable with a private_interface attribute value of "in" may be mapped to a single variable owned by a component in the encapsulated set, provided the target variable has a public_interface attribute value of "out".
> A variable with a private_interface attribute value of "out" may be mapped to variables owned by components in the encapsulated set, provided the target variables have public_interface attribute values of "in".

Reply:
this look ok, much clearer after the refactoring